### DISCUSSION TITLE
How to add external script?

### DISCUSSION THREAD

**Discussion by RidgewellServices** [epicweb-dev/epic-stack#828] [Q&A]
I've been struggling to add an external script to my project. I've tried two different platforms--mailer-lite and substack. The scripts are meant to open a newsletter signup form. I can't get past the CSP nonce errors. Is it possible to do this?

**Comment by kentcdodds** [comment]
Unless you [changed the CSP](https://github.com/epicweb-dev/epic-stack/blob/main/docs/security.md#content-security-policy) the nonce errors should just be warnings and things should work fine. But yeah you don't want to have those in there. You'll need to pass a `nonce` to any scripts you render similar to what we're doing here: https://github.com/epicweb-dev/epic-stack/blob/bd837f477fdee9aeac869b98226e2aba34104968/app/root.tsx#L183-L184 (you can put it directly on a `<script>` though. You can access the nonce in any component using the `useNonce` hook: https://github.com/epicweb-dev/epic-stack/blob/main/app/utils/nonce-provider.ts#L5

**Reply by RidgewellServices** [reply]
I've tried adding nonce. It ends up with warnings like this:
chunk-RXX2RYKY.js?v=b6341b36:521 Warning: Prop `nonce` did not match. Server: "" Client: "2b13ec4f715c4e61592411bf09caeff3"
The script won't load. I'll read about the hook to see if I can make any progress with it there.

**Reply by RidgewellServices** [reply]
It's already using the nonce-provider... completely beyond me how to make it work.

**Reply by kentcdodds** [reply]
That's odd because that's the opposite of what I would expect. The server should have a nonce and the client should not (and React is fine with this because it's how a nonce works).

https://www.epicstack.dev has a working nonce setup without warnings so there must be a difference between what you're doing and what's in the stack.

**Reply by RidgewellServices** [reply]
If I add the script to a virgin project, I get the errors. I don't have time to replicate it this afternoon, but will attack it again tomorrow. One more question - can I add:
<iframe src="https://ramonaridgewell.substack.com/embed" width="480" height="320" style="border:1px solid #EEE; background:white;" frameborder="0" scrolling="no"></iframe>

I had similar issues with this to the mailer-lite <script>.

**Reply by koikar** [reply]
I just went through the same problem for the past days. 

My learnings:

- it's not a problem with nonce itself, but actually a hydration error
- certain scripts cause this error, others don't
- Google Analytics works fine but Google Tag Manager causes hydration error (same for Segment.io and different chat widgets)

Solution:
- upgrade to React v19 canary and all hydration errors are resolved automagically

Still now I am having problems with the client side ENV variables (used by the scripts), that suddenly become undefined in production at random moments (e.g. 5 hours after deploying).

Not sure if this is caused by React v19 or enabling single_fetch. Will need to investigate more. Hopefully you don't experience this problem.

**Reply by RidgewellServices** [reply]
Thanks. Yes, there were hydration issues. I'll try the upgrade and see if that makes a difference.
